tests/unit/components/test_sa_token_component.py

```diff
@@ -0,0 +1,423 @@
+            assert kwargs["name"] == self.service_account_name
+            assert kwargs["namespace"] == self.namespace
+            spec = kwargs["body"].spec
+            assert spec.audiences == self.audiences
+            assert spec.expiration_seconds == self.expiration
+
+    def test_failing_k8s_api_irrelevant_when_not_leader(
+        self, harness_with_container, clean_service_account_token_test_directory
+    ):
+        """Check the token is not generated when the unit is not leader and the K8s API fails."""
+        # Arrange:
+
+        sa_token_dir = clean_service_account_token_test_directory
+
+        harness_with_container.set_leader(False)
+
+        harness_with_container.set_can_connect(self.container_name, True)
+        sa_token_component = SATokenComponent(
+            charm=harness_with_container.charm,
+            name=self.token_k8s_name,
+            audiences=self.audiences,
+            sa_name=self.service_account_name,
+            sa_namespace=self.namespace,
+            filename=self.token_filename,
+            path=sa_token_dir,
+            expiration=self.expiration,
+        )
+
+        with (
+            patch(K8S_CLIENT_LOAD_CONFIG_PATH),
+            patch.object(CoreV1Api, K8S_CLIENT_CREATE_TOKEN_API) as mocked_k8s_client_create_token,
+        ):
+            mocked_k8s_client_create_token.side_effect = Exception(
+                "K8s API call to generate token failed."
+            )
+
+            # Act:
+
+            sa_token_component.configure_charm("mocked event")
+
+            # Assert:
+
+            # charm status:
+            with pytest_raises(GenericCharmRuntimeError) as error:
+                sa_token_component.status
+            assert error.value.msg == (
+                f"Token file for {self.service_account_name} ServiceAccount not present in charm."
+            )
+
+            # ServiceAccount token file:
+            expected_sa_token_file_path = Path(sa_token_dir, self.token_filename)
+            assert not expected_sa_token_file_path.exists()
+
+            # K8s API calls:
+            mocked_k8s_client_create_token.assert_not_called()
+
+    @mark.parametrize("is_leader", (False, True))
+    def test_previously_created_sa_token_recognized(self, harness_with_container, is_leader):
+        """Check the previously created token file is recognized."""
+        # Arrange:
+
+        harness_with_container.set_leader(is_leader)
+
+        harness_with_container.set_can_connect(self.container_name, True)
+        sa_token_component = SATokenComponent(
+            charm=harness_with_container.charm,
+            name=self.token_k8s_name,
+            audiences=self.audiences,
+            sa_name=self.service_account_name,
+            sa_namespace=self.namespace,
+            filename=PRECREATED_SA_TOKEN_FILENAME,
+            path=PRECREATED_SA_TOKEN_DIR,
+            expiration=self.expiration,
+        )
+
+        # Act:
```

Comment:
nit: I think the call to `SATokenComponent` is the `Act` in this case so I'd put the comment above